File: c_code/src/string_const_diagram.cpp
#include "string_const_diagram.h"

/*
    See diagram.h 
*/
std::pair<DiagramProcessing,
 std::pair<std::string, std::string>> StringConstDiagram::parse(const char entry) {

    // Processing has ended.
    if (_processing_state != IN_PROGRESS) {
        return _result;
    }

    std::pair<std::string, std::string> token_and_lexem;
    std::pair<DiagramProcessing, std::pair<std::string, std::string>> return_pair;

    switch (_current_state)
    {
        case 0:
            if (entry == '\"') {
                _current_state = 1;
                _current_lexem += entry;

                token_and_lexem = {"", ""};
                return_pair.first = IN_PROGRESS;
                return_pair.second = token_and_lexem;
            } else {
                token_and_lexem = {GENERIC_TOKEN, ""};
                return_pair.first = FAILED;
                return_pair.second = token_and_lexem;
            }
            break;
        
        case 1:
            if (isalnum(entry)) {
                // _current_state = 1;
                _current_lexem += entry; 

                token_and_lexem = {"", ""};
                return_pair.first = IN_PROGRESS;
                return_pair.second = token_and_lexem;

            } else if(entry == '\"') {
                // _current_state = 2; 
                token_and_lexem = {STRING_TOKEN, _current_lexem};
                return_pair.first = FINISHED;
                return_pair.second = token_and_lexem;
            } else {
                token_and_lexem = {GENERIC_TOKEN, ""};
                return_pair.first = FAILED;
                return_pair.second = token_and_lexem;
            }
            break;
            
        default:
            assert(false);  // Should'nt run this.
            break;
    }

    _processing_state = return_pair.first;
    if (_processing_state != IN_PROGRESS) {
        _result = return_pair;
    }

    return return_pair;
}

File: compiler/lexical_analyzer/string_const_diagram.py
from .diagram import *


class StringConstDiagram(Diagram):
    def __init__(self) -> None:
        super().__init__()

    # See diagram.h
    def parse(self, entry: str) -> tuple[int, tuple[str, str]]:
        # Processing has ended.
        if self._processing_state != DiagramProcessing.IN_PROGRESS:
            return self._result

        # Inicialização de pares de token e lexema
        token_and_lexem: tuple[str, str] = ("", "")
        return_pair: tuple[int, tuple[str, str]] = (DiagramProcessing.IN_PROGRESS, ("", ""))

        # Simulação de switch-case com if-elif-else
        if self._current_state == 0:
            if entry == '"':
                self._current_state = 1
                self._current_lexem += entry

                token_and_lexem = ("", "")
                return_pair = (DiagramProcessing.IN_PROGRESS, token_and_lexem)
            else:
                token_and_lexem = (GENERIC_TOKEN, "")
                return_pair = (DiagramProcessing.FAILED, token_and_lexem)

        elif self._current_state == 1:
            if entry.isalnum():
                # self._current_state = 1
                self._current_lexem += entry

                token_and_lexem = ("", "")
                return_pair = (DiagramProcessing.IN_PROGRESS, token_and_lexem)

            elif entry == '"':
                # self._current_state = 2
                self._current_lexem += entry  # Adiciona a aspa final à string
                token_and_lexem = (STRING_TOKEN, self._current_lexem)
                return_pair = (DiagramProcessing.FINISHED, token_and_lexem)

            else:
                token_and_lexem = (GENERIC_TOKEN, "")
                return_pair = (DiagramProcessing.FAILED, token_and_lexem)

        else:
            assert False  # Shouldn't run this.

        # Atualizando estado de processamento
        self._processing_state = return_pair[0]
        if self._processing_state != DiagramProcessing.IN_PROGRESS:
            self._result = return_pair

        return return_pair